test/jdk/sun/security/pkcs11/fips/NssdbPin.java

```diff
@@ -0,0 +1,362 @@
+            p.secprop(FIPS_NSSDB_PATH_PROP, "");
+            p.secprop(FIPS_NSSDB_PIN_PROP, "");
+        } else if (ctx.propType == PropType.SECURITY) {
+            p.secprop(FIPS_NSSDB_PATH_PROP, ctx.nssdbPath);
+            pinPropVal = escapeForPropsFile(pinPropVal);
+            p.secprop(FIPS_NSSDB_PIN_PROP, pinPropVal);
+        } else {
+            throw new Exception("Unsupported property type.");
+        }
+        if (DEBUG) {
+            p.inheritIO();
+            p.prop("java.security.debug", "sunpkcs11");
+            p.debug(NssdbPin.class.getName());
+
+            // Need the launched process to connect to a debugger?
+            //System.setProperty("test.vm.opts", "-Xdebug -Xrunjdwp:" +
+            //         "transport=dt_socket,address=localhost:8000,suspend=y");
+        } else {
+            p.nodump();
+        }
+        procCb.accept(p);
+        p.start().waitFor(0);
+    }
+
+    private static String escapeForPropsFile(String str) throws Throwable {
+        StringBuffer sb = new StringBuffer();
+        for (int i = 0; i < str.length(); i++) {
+            int cp = str.codePointAt(i);
+            if (Character.UnicodeBlock.of(cp)
+                    == Character.UnicodeBlock.BASIC_LATIN) {
+                sb.append(Character.toChars(cp));
+            } else {
+                sb.append("\\u").append(String.format("%04X", cp));
+            }
+        }
+        return sb.toString();
+    }
+
+    private static void createNSSDB(TestContext ctx) throws Throwable {
+        ProcessBuilder pb = getModutilPB(ctx, Arrays.asList("-create"), null);
+        if (DEBUG) {
+            System.out.println("Creating an NSS DB in " + ctx.workspace +
+                    "...");
+            System.out.println("cmd: " + String.join(" ", pb.command()));
+        }
+        if (pb.start().waitFor() != 0) {
+            throw new Exception("NSS DB creation failed.");
+        }
+        generatePinFile(ctx);
+        pb = getModutilPB(ctx, Arrays.asList("-changepw", NSSDB_TOKEN_NAME),
+                Arrays.asList("-newpwfile", ctx.nssdbPinFile.toString()));
+        if (DEBUG) {
+            System.out.println("NSS DB created.");
+            System.out.println("Changing NSS DB PIN...");
+            System.out.println("cmd: " + String.join(" ", pb.command()));
+        }
+        if (pb.start().waitFor() != 0) {
+            throw new Exception("NSS DB PIN change failed.");
+        }
+        if (DEBUG) {
+            System.out.println("NSS DB PIN changed.");
+        }
+    }
+
+    private static ProcessBuilder getModutilPB(TestContext ctx,
+            List<String> opts, List<String> args) throws Throwable {
+        ProcessBuilder pb = new ProcessBuilder("modutil", "-force");
+        List<String> pbCommand = pb.command();
+        if (opts != null) {
+            pbCommand.addAll(opts);
+        }
+        if (args != null) {
+            pbCommand.addAll(args);
+        }
```

Comment:
The distinction between arguments and options matters (and is documented) at the _modutil_ level because of the order: you cannot pass an argument before an option. I'll leave this comment unresolved in case that you want to add something else to this conversation.

Reply:
Ah ok, so the semantic is in `modutil` itself. If I only pay attention to the _Java_ code, I still think that having `getModutilPB()` calls like the following would read similar enough to a `modutil` command in the terminal, to make them self explanatory.
```java
pb = getModutilPB(ctx, "-create");
//--> modutil -force -create -dbdir <nssdbPath>

pb = getModutilPB(ctx, "-changepw", NSSDB_TOKEN_NAME, "-newpwfile", ctx.nssdbPinFile.toString());
//--> modutil -force -changepw <NSSDB_TOKEN_NAME> -newpwfile <nssdbPinFile> -dbdir <nssdbPath>
```


Also, since the current code doesn't handle them in any particular way, we can have the following:
```java
// Right
ProcessBuilder pb = getModutilPB(ctx, Arrays.asList("-create"), null);

// Semantically wrong, but works exactly the same
ProcessBuilder pb = getModutilPB(ctx, null, Arrays.asList("-create"));
```

I just want to make my point clear, but I repeat that none of this blocks the review approval though.

Reply:
Yes, I see your point. We can make this change if the function getModutilPB only adds options at the beginning and arguments at the end, and if the caller gives an ordered list. Will implement the change.